### ISSUE TITLE
🦠 Model Request: Tox21-MMP

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#459]
### Model Title

Tox21-MMP (TDCommons)

### Publication

Hello @Celinezzzzz !

As part of your Outreachy contribution, we have assigned you the dataset "Tox21-MMP" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by GemmaTuron** [comment]
Hello @Celinezzzzz

Hope you are doing fine,please note that today was the last day to try out the models and explain the results you are getting, I'd suggest:

- Finish up your contribution and post the results here
- Close the issue
- Link this issue to your outreachy contribution profile
- Move on to prepare final application

**Comment by Celinezzzzz** [comment]
**Dataset Overview: Tox21-SR-MMP**
Tox21 is a data challenge which contains qualitative toxicity measurements for 7,831 compounds on 12 different targets, such as nuclear receptors and stree response pathways.

**Access to the Google Colab Notebook:** 
https://colab.research.google.com/drive/1qJaGNd2IgK9LrmKKr3nT2ybCvvvFOHYv?usp=sharing

**Steps to Build the Model:**
1.	Loading the data from Therapeutics Data Commons
2.	Data Analysis and Visualization
3.	Scaling the data and covert the labels for the classification task
4.	Data Preprocessing and Model Training using LazyQSAR
5.	Evaluating Model using Test Data and Validation Data

**Loading the data from Therapeutics Data Commons**
Loaded the Tox21 Dataset from Therapeutics Data Commons. SR-MMP is one of the assays among Tox21 datasets, which was retrieved by specifying the label name in the label_name variable to the data loader.

**What are we trying to predict?**
We are trying to predict the toxicity of a drug SMILES string using a binary classifier. Tox21-SR-MMP consists of 5810 drugs (Drug ID and SMILES).

**Data Analysis and Visualization**
We divide the loaded data between train/validation/test sets which are already split through the pyTDC module using data.get_split(). Column Drug represents the molecules in Simplified Molecular Input Line Entry System (SMILES), a chemical notation for chemical structure that allows the computer to understand. Column Y represents values 0 or 1 a classification for active(toxic) 1 or inactive(non-toxic) 0. This is useful as this is a Binary Classification Model.

![1667575698361](https://user-images.githubusercontent.com/99721409/200013885-ee387f33-83a2-40a7-be68-9040648d5869.png)

**How many datapoints do we have in total and in each set (train / validation / test)?** 
Total Dataset: 5810:
Train set: 4067
Validation set: 581
Test set: 1162
 
**How many actives and inactives do we have?** 
Train dataset
Number of inactive: 3439
Number of active: 628

Validation dataset
Number of inactive: 472
Number of active: 109

Test dataset
Number of inactive: 981
Number of active: 181

![1667575326961](https://user-images.githubusercontent.com/99721409/200011817-af3acbd5-d102-4c6f-9ffb-eebd746cb40e.png)


**Can you show in a plot the distribution of actives/inactives, or the values for the regression?**
I visualized the data active/inactive distribution through building a pie chart and bar chart using matplotlib package
![1667575364901](https://user-images.githubusercontent.com/99721409/200011953-702e2686-061e-4882-af89-1e694a86eade.png)
![1667575394371](https://user-images.githubusercontent.com/99721409/200012004-23f315d2-5058-4864-a19e-ce93b579c6b4.png)
![1667575410225](https://user-images.githubusercontent.com/99721409/200012063-60c12c25-df59-4e7e-85bd-a9d323038d46.png)

**Draw Molecular structure of a toxic and intoxic molecule using the RDKIT package**

![1667575444614](https://user-images.githubusercontent.com/99721409/200012277-5f3a5434-73b3-41df-8465-864ec12befbf.png)
![1667575458230](https://user-images.githubusercontent.com/99721409/200012448-cd3aea6a-22ef-4c11-bdd6-6f07db307095.png)

**Data Preprocessing and Model Training using LazyQSAR**
The input data from the other model is the drug's SMILE structure. For a machine learning model, the input data needs to be passed into a numeric/vector format. Lazy-qsar package automatically converts SMILES to Morgan fingerprints which can be used to fit the data into Morgan Binary Classifier. We use a classification model because it is a classification problem, this is so due to the fact that the outcome in the dataset (Y) is either 0 (inactive) 1 (active) which is a discrete outcome. When the outcome is discrete , it is a classification problem

Model Training Time: 11 seconds

**Evaluating Model using Test Data and Validation Data**

**Create a ROC Curve with the Validation Set**
Metrices used to measure the performance of the model are **ROC Curve** and **Confusion Matrix.**

**ROC Curve with Validation Set**
![1667575508927](https://user-images.githubusercontent.com/99721409/200012940-2d76b563-cce2-4456-8783-e7aee14edaec.png)

**How do you interpret the ROC Curve you got?**
An ROC curve (receiver operating characteristic curve) is a graph showing the performance of a classification model at all classification thresholds. This curve plots two parameters:

True Positive Rate False Positive Rate True Positive Rate (TPR) is a synonym for recall and is therefore defined as the number of true positives divided by the number of actual positives (TP + FN). False Positive Rate (FPR) is defined as FN / (FP + TN).

From the curve above it shows that the probability that the model will assign a larger probability to a random positive example than a random negative example is 0.8756802985538796.

This ROC curve has an AUC of 0.88 which is greater than 0.5, meaning it ranks a random positive example higher than a random negative example more than 50% of the time.

**Confusion Matrix with Validation Set**
![1667575527107](https://user-images.githubusercontent.com/99721409/200013257-d77ed7b8-be4a-4e71-b7b8-25cb4a2ebd66.png)

Precision: 0.5041322314049587 
Recall: 0.5596330275229358 
f1_score: 0.5304347826086957

The precision is intuitively the ability of the classifier not to label as positive a sample that is negative.
The recall is intuitively the ability of the classifier to find all the positive samples.

**Evaluation on Test Set**
![1667575565199](https://user-images.githubusercontent.com/99721409/200013380-f87aac7b-1608-41d1-9202-c25422da2ee4.png)

![1667575580866](https://user-images.githubusercontent.com/99721409/200013442-5ec2661f-9ab3-439c-807d-47918758e875.png)

Precision: 0.3870967741935484 
Recall: 0.46408839779005523 
f1_score: 0.42211055276381904

**What is the performance of your model? What metrics are you using and why?**
The high value for AUC_ROC for the test i.e. 0.74 indicates the model to be good but the Precision and Recall scores are low.

**Comment by GemmaTuron** [comment]
Hello @Celinezzzzz !

Thanks for your contributions to Ersilia during the Outreachy Application Period! We hope you have learnt and enjoyed as much as the Ersilia team did!
I will close this issue as this was part of the application period solely.
